### PR TITLE
Basics ~to curve fitting~

### PR DESCRIPTION
**Pull request by Vindaar**
I starting writing a chapter that is supposed to take people from "what types are there in Nim" to fitting a function and plotting it.

I somehow messed up my priorities though and ended up writing too much basic stuff so I ran out of time for now. I don't think what I wrote is particularly great, but well. I'll put it up as a draft in case someone wants to give it a read.

edit: instead of covering all of the above in this PR, I'll stop after the most basic things. With these in place there's at least some content and a bit of structure.

### CODE REVIEW

**Review by HugoGranstrom — COMMENTED** [review]
Had a read through the finished parts and they are so thorough :) I made a few comments with things to discuss. 

All of the parts in this `road to fit` seem pretty standalone to me and especially the `datatypes` and `data wrangling` seem like quite general topics which could belong in an `Introduction` section or something like that. If one still wanted to keep the feeling of this "road to fit" one could add a `Prerequisites` section at the top of the `Curve Fitting` article listing the previous articles. What do you think? 😄

**Comment by HugoGranstrom** [comment]
CI is failing because the packages like `numericalnim` and `datamancer` and `mpfit` isn't in the nimble file btw.

**Comment by Vindaar** [comment]
> Had a read through the finished parts and they are so thorough :) I made a few comments with things to discuss.

Thanks a lot!

> All of the parts in this `road to fit` seem pretty standalone to me and especially the `datatypes` and `data wrangling` seem like quite general topics which could belong in an `Introduction` section or something like that. If one still wanted to keep the feeling of this "road to fit" one could add a `Prerequisites` section at the top of the `Curve Fitting` article listing the previous articles. What do you think?

Oh, yes I suppose they are. I simply thought about writing something about using data frames and fitting and realized that without some kind of introductory basics, that's a bit out of place. So I thought up this structure here.

But we can totally turn it into a more general structure. Pretty much what you suggest. Make data types and data wrangling their own things under "introduction". Then in a different section have the fitting with a "prerequisites: read these before" thing. Plotting can go in either place really.

File: book/road_to_fit/common_datatypes.nim [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+"""
+nbCode:
+  y2[0] = 5.0
+  echo y2
+nbText: """
+
+New elements are added using `add` as is typical in Nim:
+"""
+nbCode:
+  y1.add 10.0
```
[the remark this answers is not in the document]
Comment (Vindaar) [reply book/road_to_fit/common_datatypes.nim:94]: Ok, good point. I'll change that.

File: book/road_to_fit/common_datatypes.nim [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+  echo "Length: ", y1.len
+nbText: """
+So `y1` now contains 1 element instead of 0.
+
+Deleting elements is also supported, via `delete` or `del`. Both procedures take the index
+to be removed. `delete` keeps the order of the sequence intact, whereas `del` simply overwrites
+the given index with the last element of the sequence and reduces the length by one. Compare:
+"""
+nbCode:
+  var x2 = x1
```
[the remark this answers is not in the document]
Comment (Vindaar) [reply book/road_to_fit/common_datatypes.nim:105]: Same. :+1:

File: book/road_to_fit/common_datatypes.nim [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+nbText: """
+First we can just create a tensor from a (possibly nested) sequence or array using `toTensor`.
+
+Secondly:
+"""
+nbCode:
+  let t2 = newTensor[float](9)
+nbText: """
+This is the default tensor constructor. It creates a tensor of type `Tensor[float]` with
+10 elements that is zero initialized. If multiple elements are given to the procedure a tensor
```
[the remark this answers is not in the document]
Comment (Vindaar) [reply book/road_to_fit/common_datatypes.nim:248]: Haha, indeed. I had a 10 there first, but then thought with 9 I can do a reshape to `[3, 3]` and forgot to change the text afterwards.

**Comment by HugoGranstrom** [comment]
> Oh, yes I suppose they are. I simply thought about writing something about using data frames and fitting and realized that without some kind of introductory basics, that's a bit out of place. So I thought up this structure here.

You have paved the way for everyone else now so no one else after you have to do it as well :D

> But we can totally turn it into a more general structure. Pretty much what you suggest. Make data types and data wrangling their own things under "introduction". Then in a different section have the fitting with a "prerequisites: read these before" thing. Plotting can go in either place really.

Yeah, the structure part is really fiddly right now how the structure will end up. I think we have some good coverage in #19 so we can start sketching on what the broadest topics should be. What topic do you think `Curve Fitting` would fit under? Feels like it would belong in the same category as interpolation and things like that. And the `Plotting` would either be in `Introduction` or `Data Visualization` I guess? 

I imagine something along these lines:
```
Introduction/
	Datatypes
	Data Wrangling Basics
	(Plotting Basics?)
Interpolation and Curve Fitting/
	Curve Fitting Basics
	Interpolation Basics
	...
Data Visualization/
	(Plotting Basics?)
	Other plotting in-depth tutorials
	...
```

**Comment by Vindaar** [comment]
Applied the changes you mentioned and continued on a bit on the DF stuff.

I like your proposed structure. The "Interpolation and Curve Fitting" section could be a general "numerical methods" section or something like this.

**Comment by HugoGranstrom** [comment]
Thanks! Nice, will take another read later tonight :)

> I like your proposed structure. The "Interpolation and Curve Fitting" section could be a general "numerical methods" section or something like this.

Yes, `Numerical Methods` sounds even better! :D

**Comment by HugoGranstrom** [comment]
Btw, it would be nice if we could have a "preview deploy" of PRs so one can read through PRs without having to locally build them. (And we make sure that the built version works only, https://github.com/pietroppeter/nimibook/issues/30 shows that things can work locally and break when deployed)

**Comment by Vindaar** [comment]
> Btw, it would be nice if we could have a "preview deploy" of PRs so one can read through PRs without having to locally build them. (And we make sure that the built version works online, [pietroppeter/nimibook#30](https://github.com/pietroppeter/nimibook/issues/30) shows that things can work locally and break when deployed)

Yeah, I totally agree. But I don't really see an easy way to accomplish that. It needs to be hosted somewhere.

**Comment by HugoGranstrom** [comment]
> Yeah, I totally agree. But I don't really see an easy way to accomplish that. It needs to be hosted somewhere.

I think for another (non-Nim) project I'm a collab of the guy getting stuff done there used Netlify and specifically [Deploy Previews](https://www.netlify.com/blog/2016/07/20/introducing-deploy-previews-in-netlify/). Haven't done much research on it though but it looks like it may be possible to use with their free tier.

**Comment by Vindaar** [comment]
It doesn't build btw, because of an ambiguous identifier issue between `Value` from datamancer and `Value` from mustache. 

I tried to change the datamancer code to fully qualify the symbol, but on Nim 1.4 that fails for some reason. So either we generate the book on devel, or I'm a bit out of ideas right now.

**Comment by HugoGranstrom** [comment]
That's odd :/ I'm on devel locally so didn't notice 🙃

**Comment by HugoGranstrom** [comment]
I've read through it now and the majority of it is clear as crystal water! Good job! :D There's really just one section I had problems wrapping my head around at first. And it was the long vs wide format. Had never heard of it before so I would have appreciated showing an example of a wide and a long df at the beginning of the section (it was shown both above though but with additional columns, and below but I hadn't read that yet when my brain got caught on this 🙃). 

Also, I didn't see `col` or `idx` mentioned anywhere, is that a conscious decision?

**Comment by Vindaar** [comment]
Thanks for the feedback!

I'll change the long / wide paragraph a little bit to make it clearer.

> Also, I didn't see col or idx mentioned anywhere, is that a conscious decision?

Yes, that was a conscious decision. I didn't want to make it more complicated than it needs to be. For most use cases just referring to columns using accented quotes is perfectly enough. There can either be a more complex data frame tutorial at some point or it's enough to reference the documentation (once that is actually well explained there).

**Comment by Vindaar** [comment]
Ok, the paragraph about `gather` and long vs. wide format was extended. Feel free to re-read that part. :)

**Review by HugoGranstrom — APPROVED** [review]

**Comment by HugoGranstrom** [comment]
That made such a big difference in understanding the long vs wide, thank you for taking your time and clarifying it. I'm content with merging this now 😃

**Pull request merged**